midend/local_copyprop.cpp

```diff
@@ -152,9 +177,33 @@ IR::P4Action *P4::DoLocalCopyPropagation::postorder(IR::P4Action *act) {
     LOG5("DoLocalCopyPropagation before ElimDead " << act->name);
     LOG5(act);
     act->body = act->body->apply(ElimDead(*this))->to<IR::BlockStatement>();
-    in_action = false;
+    working = false;
     available.clear();
     LOG3("DoLocalCopyPropagation finished action " << act->name);
     LOG4(act);
     return act;
 }
+
+IR::P4Control *P4::DoLocalCopyPropagation::preorder(IR::P4Control *ctrl) {
+    if (working || !available.empty()) BUG("corrupt internal data struct");
+    visit(ctrl->type, "type");
+    visit(ctrl->constructorParams, "constructorParams");
+    visit(ctrl->controlLocals, "controlLocals");
+    if (working || !available.empty()) BUG("corrupt internal data struct");
+    working = true;
+    LOG2("DoLocalCopyPropagation working on control " << ctrl->name);
+    LOG4(ctrl);
+    for (auto local : *ctrl->controlLocals)
+        if (auto var = local->to<IR::Declaration_Variable>())
+            visit_local_decl(var);
+    visit(ctrl->body, "body");
+    LOG5("DoLocalCopyPropagation before ElimDead " << ctrl->name);
+    LOG5(ctrl);
+    ctrl->body = ctrl->body->apply(ElimDead(*this))->to<IR::BlockStatement>();
```

Comment:
As named, this is just LOCAL copyprop -- it only looks at the body of the control.  So it is unable to eliminate vars if the control has any table applies, but does simplify expressions that will later get converted into tables (which is worthwhile).  I suspect we will need to look at the effects of actions from table.apply calls to allow better cleanup, which is non-local, but shouldn't be too expensive (as long as said actions don't call other actions -- may need to build a global table of what each action accesses, and then do closure based on calls)